arviz/kde_utils.py

```diff
@@ -0,0 +1,895 @@
+
+
+def kappa_mle(x):
+    mean = circular_mean(x)
+    kappa = a1inv(np.mean(np.cos(x - mean)))
+    return kappa
+
+
+def dct1d(x):
+    """
+    Discrete Cosine Transform in 1 Dimension
+
+    Parameters
+    ----------
+    x : numpy array
+        1 dimensional array of values for which the
+        DCT is desired
+
+    Returns
+    -------
+    output : DTC transformed values
+    """
+
+    x_len = len(x)
+
+    even_increasing = np.arange(0, x_len, 2)
+    odd_decreasing = np.arange(x_len - 1, 0, -2)
+
+    x = np.concatenate((x[even_increasing], x[odd_decreasing]))
+
+    w_1k = np.r_[
+        1,
+        (2 * np.exp(-(0 + 1j) * (np.arange(1, x_len)) * np.pi / (2 * x_len)))
+    ]
+    output = np.real(w_1k * fft(x))
+
+    return output
+
+
+def _fixed_point(t, N, k_sq, a_sq):
+    """
+    Implementation of the function t-zeta*gamma^[l](t) derived
+    from equation (30) in [1]
+
+    References
+    ----------
+    .. [1] Kernel density estimation via diffusion.
+       Z. I. Botev, J. F. Grotowski, and D. P. Kroese.
+       Ann. Statist. 38 (2010), no. 5, 2916--2957.
+    """
+
+    k_sq = np.asfarray(k_sq, dtype=np.float64)
+    a_sq = np.asfarray(a_sq, dtype=np.float64)
+
+    l = 7
+    f = np.sum(np.power(k_sq, l) * a_sq * np.exp(-k_sq * np.pi ** 2 * t))
+    f *= 0.5 * np.pi ** (2.0 * l)
+
+    for j in reversed(range(2, l)):
+        c1 = (1 + 0.5 ** (j + 0.5)) / 3
+        c2 = np.product(np.arange(1.0, 2 * j + 1, 2, dtype=np.float64))
+        c2 /= (np.pi / 2) ** 0.5
+        t_j = np.power((c1 * (c2 / (N * f))), (2.0 / (3.0 + 2.0 * j)))
+        f = np.sum(k_sq ** j * a_sq * np.exp(-k_sq * np.pi ** 2.0 * t_j))
+        f *= 0.5 * np.pi ** (2 * j)
+
+    out = t - (2 * N * np.pi ** 0.5 * f) ** (-0.4)
+    return out
+
+
+def _root(function, N, args, x):
+    # The idea here was borrowed from KDEpy implementation.
+    # The right bound is at most 0.01
+    found = 0
+    N = max(min(1050, N), 50)
+    tol = 10e-12 + 0.01 * (N - 50) / 1000
+
+    while found == 0:
+        try:
+            bw, res = brentq(function, 0, 0.01, args=args, full_output=True,
+                             disp=False)
+            found = 1 if res.converged else 0
+        except ValueError:
+            bw = 0
+            tol *= 2.0
+            found = 0
+        if bw <= 0:
+            warnings.warn(
+                "Improved Sheather-Jones did not converge to a positive value. "
+                "Using Silverman's rule instead.",
+                Warning
+            )
+            bw = (bw_silverman(x) / np.ptp(x)) ** 2
+            return bw
+        if tol >= 1:
+            warnings.warn(
+                "Improved Sheather-Jones did not converge. "
+                "Using Silverman's rule instead.",
+                Warning
+            )
+            bw = (bw_silverman(x) / np.ptp(x)) ** 2
+    return bw
+
+# KDE Utilities ---------------------------------------------------------------------
+
+def check_type(x):
```

Comment:
I am not sure about what differences in behaviour has this versus some `np.asarray().flatten()` or `np.array().flatten()` (or maybe atleast_1d is needed?), I understand we want to make sure we get a numpy array and that the array is 1d right?

Reply:
My concern with `np.asarray().flatten()` is that it works for something like `l = [[1, 2], [3, 4]]` but does not work for `l = [[1, 2], [3, 4, 5]]`.

```python
l = [[1, 2], [3, 4]]
np.array(l).flatten()
# array([1, 2, 3, 4])

l = [[1, 2], [3, 4, 5]]
np.array(l).flatten()
# array([list([1, 2]), list([3, 4, 5])], dtype=object)
```

With "works" I mean to return a 1d array of numbers. In the second case it is a 1d array but made of lists.
If I do `np.asfarray(x).flatten()` it works correctly in the first case and raises an error in the second.

I have the following ideas:

* Call `x = np.asfarray(x).flatten()` instead of the current `x = np.asfarray(x)` plus dimension number checking. If dimension of `x` is larger than 1 two things can happen. It is either successfully and silently flattened or an error is raised. 
* Call `x = np.asfarray(x)`. If dimension of `x` is larger than 1D and similar to the second example, it will raise an error. Otherwise it won't. Then check dimension number. If this is larger than 1, call `flatten()` method and raise a warning to inform the user the original object `x` was not 1-dimensional but was successfully converted. 
* Like the previous idea but without raising any warning if successfully flattening.

Personally, I'm more towards being more informative and signalling error/warnings. I think that if the user accidentally passes an object whose dimension is larger than 1 we should not resolve the situation silently - we can't be sure this is his/her original intention.
However, since this is my first contribution to the project and there may be other internal conventions that I'm not aware of,  I think it is good to follow your advise in this issue. Please let me know what would you do here.

Reply:
I think the second case raising an error is to be expected, and silently flattening is perfectly ok and coherent with the rest of ArviZ. 

`kde` will generally not be called by users, they will use plot_kde, plot_posterior... I think that kde, like plot_kde and plot_dist who also take an array input, is expected to generate a single kde even if the input is multidimensional (i.e. scalar variables will generally be stored as 2d arrays with (chain, draw) dims). To get multiple kdes, plot_posterior, plot_density... should be used